src/main/webapp/resources/js/pages/projects/samples-metadata-import/redux/importReducer.ts

```diff
@@ -91,79 +95,173 @@ export const saveMetadata = createAsyncThunk<
-              createSample({
-                projectId,
-                body: {
-                  name,
-                  metadata: metadataFields,
-                },
-              })
-                .then(() => {
-                  metadataSaveDetails[index] = { saved: true };
-                })
-                .catch((error) => {
-                  metadataSaveDetails[index] = {
-                    saved: false,
-                    error: error.response.data.error,
-                  };
-                })
-            );
-          }
-        }
-      }
-      await Promise.all(promises).then(() => {
-        dispatch(
-          setMetadataSaveDetails(Object.assign({}, metadataSaveDetails))
+    const selectedSampleList = metadata.filter((metadataItem) => {
+      const name: string = metadataItem[sampleNameColumn];
+      return (
+        selectedMetadataKeys.includes(metadataItem.rowKey) &&
+        metadataSaveDetails[name]?.saved !== true
+      );
+    });
+
+    //update existing project samples
+    const updateSamplesPromises: Promise<void>[] = [];
+    const updateSampleList = selectedSampleList
+      .filter((metadataItem) => {
+        const name = metadataItem[sampleNameColumn];
+        const sampleId = metadataValidateDetails[name].foundSampleId;
+        return sampleId;
+      })
+      .map((metadataItem) => {
+        const name = metadataItem[sampleNameColumn];
+        const sampleId = metadataValidateDetails[name].foundSampleId;
+        const metadataFields = Object.entries(metadataItem)
+          .filter(
+            ([key]) =>
+              headers.map((header) => header.name).includes(key) &&
+              key !== sampleNameColumn
+          )
+          .map(([key, value]) => ({
+            field: key,
+            value,
+            restriction: headers.filter((header) => header.name === key)[0]
+              .targetRestriction,
+          }));
+        return { name, sampleId, metadata: metadataFields };
+      });
+    //create a request with a list of project samples to be updated
+    if (updateSampleList.length > 0) {
+      const updateSampleListChunkSize = calculateChunkSize(
+        updateSampleList.length
+      );
+      const chunkedUpdateSampleList = chunkArray(
+        updateSampleList,
+        updateSampleListChunkSize
+      );
+      for (const chunk of chunkedUpdateSampleList) {
+        updateSamplesPromises.push(
+          updateSamples({
+            projectId,
+            body: chunk,
+          })
+            .then((response) => {
+              const { responses } = response.data;
+              Object.keys(responses).map((key) => {
+                const { error, errorMessage } = responses[key];
+                metadataSaveDetails[key] = {
+                  saved: !error,
+                  error: errorMessage,
+                };
+              });
+            })
+            .catch((error) => {
+              const { responses } = error.response.data;
```

Comment:
Can this be refactored into a function since it is duplicated on line 239.

Reply:
Please see 687f4bf19a4bf9dae4774d3137f8b19fcce41d53.